internal/pkg/template/diff/diff.go

```diff
@@ -19,10 +21,107 @@ type Node struct {
 
 // Parse constructs a diff tree that represent the differences between two YAML documents.
 func Parse(curr, old []byte) (*Node, error) {
-	return &Node{}, nil
+	var currNode, oldNode yaml.Node
+	if err := yaml.Unmarshal(curr, &currNode); err != nil {
+		return nil, fmt.Errorf("unmarshal current template: %w", err)
+	}
+	if err := yaml.Unmarshal(old, &oldNode); err != nil {
+		return nil, fmt.Errorf("unmarshal old template: %w", err)
+	}
+	return parse(&currNode, &oldNode, "")
+
 }
 
 // String returns the string representation of the tree stemmed from the diffNode n.
 func (n *Node) String() string {
 	return ""
 }
+
+func parse(curr, old *yaml.Node, key string) (*Node, error) {
+	if curr == nil || old == nil {
+		return &Node{
+			key:      key,
+			oldValue: old,
+			newValue: curr,
+		}, nil
+	}
+	if isYAMLLeaf(curr) && isYAMLLeaf(old) {
+		if curr.Value == old.Value {
+			return nil, nil
+		}
+		return &Node{
+			key:      key,
+			newValue: curr,
+			oldValue: old,
+		}, nil
+	}
+	var children map[string]*Node
+	var err error
+	switch {
+	case curr.Kind == yaml.SequenceNode && old.Kind == yaml.SequenceNode:
+		children, err = parseSequence(curr, old)
+	case curr.Kind == yaml.DocumentNode && old.Kind == yaml.DocumentNode:
+		fallthrough
+	case curr.Kind == yaml.MappingNode && old.Kind == yaml.MappingNode:
+		children, err = parseMap(curr, old)
+	default:
+		return nil, fmt.Errorf("unknown combination of node kinds: %v, %v", curr.Kind, old.Kind)
```

Comment:
Ah I am guessing this is why you've asked a while back if it's guaranteed that the API with CloudFormation never supports composite types.

Now that I'm thinking about it, I believe an IAM policy is possible counter example:
```yaml
resources: '*' # same for actions

# ----  new.yaml
resources:
  - arn1
  - arn2
```
In this scenario, I wonder if we can showcase a diff like this 💭 ?
```
~ resources: "*" -> [arn1, arn2] 
```

Reply:
Good call! 

It is possible that the item in the list is more complex than just a string.
```
Resources:
  - !Join ["", ["arn::aws::s3", "fancy"]]
  - !Join ["", ["arn::aws::s3", "extra-fancy"]]
```

With the proposal, it'd become
```
~ resources: "*" -> [!Join ["", ["arn::aws::s3", "fancy"]], !Join ["", ["arn::aws::s3", "extra-fancy"]]]
```

which can rendered to long and unreadable if the list item becomes more complex.

What do you think of this:

```
- Resources: "*" 
+ Resources:
  + - !Join ["", ["arn::aws::s3", "fancy"]]
  + - !Join ["", ["arn::aws::s3", "extra-fancy"]]
```

Reply:
Looks great, :shipit:

Reply:
Similarly, for a field that can be a scalar or map:
```
- Resources: "*" 
+ Resources:
  + Condition: condition1
  + ExceptFor: this
```

I don't know if there is any actual CFN field that is a string-or-map, but I'm planning to add the logic just in case 😂

Reply:
I plan to add this logic to a separate PR if that's alright with you!

Reply:
sounds good! I'll leave the comment as unresolved so that others can also see.